ledger/complete/mtrie/trie/trie.go

```diff
@@ -100,17 +100,27 @@ func (mt *MTrie) String() string {
 	return trieStr + mt.root.FmtStr("", "")
 }
 
-// UnsafeRead read payloads for the given paths. It is called unsafe as it modifies the
-// input paths (paths are sorted when the function returns).
+// UnsafeRead read payloads for the given paths.
+// UNSAFE: requires _all_ paths to have a length of mt.Height bits.
+// CAUTION: while reading the payloads, `paths` is permuted IN-PLACE for optimized processing.
+// Return:
+//  * `payloads` []*ledger.Payload
+//     For each path, the corresponding payload is written into payloads. AFTER
+//     the read operation completes, the order of `path` and `payloads` are such that
+//     for `path[i]` the corresponding register value is referenced by 0`payloads[i]`.
 // TODO move consistency checks from Forest into Trie to obtain a safe, self-contained API
 func (mt *MTrie) UnsafeRead(paths []ledger.Path) []*ledger.Payload {
-	// allocate the result buffer
-	res := make([]*ledger.Payload, len(paths))
-	mt.read(&res, mt.root, paths)
-	return res
+	payloads := make([]*ledger.Payload, len(paths)) // pre-allocate slice for the result
+	mt.read(mt.root, paths, payloads)
+	return payloads
 }
 
-func (mt *MTrie) read(res *[]*ledger.Payload, head *node.Node, paths []ledger.Path) {
+// read reads all the registers in subtree with `head` as root node. For each
+// `path[i]`, the corresponding payload is written into `payloads[i]` for the same index `i`.
+// CAUTION:
+//  * while reading the payloads, `paths` is permuted IN-PLACE for optimized processing.
+//  * unchecked requirement: all paths must go through the `head` node
+func (mt *MTrie) read(head *node.Node, paths []ledger.Path, payloads []*ledger.Payload) {
```

Comment:
Ok, makes sense. How about
```golang
read(payloads []*ledger.Payload, paths []ledger.Path, head *node.Node)
```
because `paths` is essentially an output too

Reply:
You're right, paths are also an output because they are sorted. 
I think this is not really important, it's fine for me to keep the APIs you suggested.

Reply:
updated method signature in https://github.com/onflow/flow-go/pull/562/commits/628dbf1df7dead82fdc9f0096c8c95a615708e4b